### PR TITLE
opcua-ha-add-log-trace-for-investigation

### PR DESCRIPTION
**Pull request by marouanehassanioptimistik**

### CODE REVIEW

**Review by jfhenon — APPROVED** [review]

**Review by burgerni10 — APPROVED** [review]

File: deps/win/HdaAgent/HdaAgent.application [hunk 1]
```diff
@@ -1,10 +1,10 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <asmv1:assembly xsi:schemaLocation="urn:schemas-microsoft-com:asm.v1 assembly.adaptive.xsd" manifestVersion="1.0" xmlns:asmv1="urn:schemas-microsoft-com:asm.v1" xmlns="urn:schemas-microsoft-com:asm.v2" xmlns:asmv2="urn:schemas-microsoft-com:asm.v2" xmlns:xrml="urn:mpeg:mpeg21:2003:01-REL-R-NS" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:asmv3="urn:schemas-microsoft-com:asm.v3" xmlns:dsig="http://www.w3.org/2000/09/xmldsig#" xmlns:co.v1="urn:schemas-microsoft-com:clickonce.v1" xmlns:co.v2="urn:schemas-microsoft-com:clickonce.v2">
   <assemblyIdentity name="HdaAgent.application" version="1.0.0.1" publicKeyToken="0000000000000000" language="neutral" processorArchitecture="msil" xmlns="urn:schemas-microsoft-com:asm.v1" />
-  <description asmv2:publisher="HdaAgent" asmv2:product="HdaAgent" xmlns="urn:schemas-microsoft-com:asm.v1" />
```
Comment (burgerni10) [inline deps/win/HdaAgent/HdaAgent.application:4]: @jfhenon Can you change that to Optimistik ?

File: src/south/ProtocolHandler.class.js [hunk 2]
```diff
@@ -266,24 +270,31 @@ class ProtocolHandler {
...
+          await this.historyQueryHandler(scanMode)
+        }
+      } catch (error) {
+        this.logger.error(`${this.dataSource.name} on scan error: ${error}.`)
       }
-    } catch (error) {
-      this.logger.error(`${this.constructor.name} on scan error: ${error}.`)
+      this.currentlyOnScan[scanMode] = 0
+    } else {
+      // eslint-disable-next-line max-len
```
Comment (burgerni10) [inline src/south/ProtocolHandler.class.js:294]: Move this line to avoid linting errors

**Review by burgerni10 — CHANGES_REQUESTED** [review]
@jfhenon  File deps/win/HdaAgent/app.publish/HdaAgent.exe can be deleted since it is not used and duplicated from deps/win/HdaAgent/HdaAgent.exe

File: src/south/OPCUA/OPCUA_HA/OPCUA_HA.class.js [hunk 3]
```diff
@@ -115,13 +118,19 @@ class OPCUA_HA extends ProtocolHandler {
...
-    this.addValues(values)
-    this.lastCompletedAt[scanMode] = new Date(maxTimestamp + 1)
+    if (values.length > 0) {
+      // send the packet immediately to the engine
+      this.addValues(values)
+      this.lastCompletedAt[scanMode] = new Date(maxTimestamp + 1)
+      await this.setConfig(`lastCompletedAt-${scanMode}`, this.lastCompletedAt[scanMode].toISOString())
+      this.logger.debug(`Updated lastCompletedAt for ${scanMode} to ${this.lastCompletedAt[scanMode]}`)
+    } else {
+      this.logger.trace(`No value for ${scanMode}`)
```
Comment (jfhenon) [inline src/south/OPCUA/OPCUA_HA/OPCUA_HA.class.js:128]: logger.warn?

File: src/south/OPCUA/OPCUA_HA/OPCUA_HA.class.js [hunk 4]
```diff
@@ -101,6 +101,9 @@ class OPCUA_HA extends ProtocolHandler {
         const selectedTimestamp = value.sourceTimestamp ?? value.serverTimestamp
         return selectedTimestamp.getTime() >= opcStartTime.getTime()
       })
+      if (newerValues.length < dataValue.length) {
+        this.logger.trace(`Received ${newerValues.length} new values for ${nodesToRead[i].pointId} among ${dataValue.length} values retrieved`)
```
Comment (jfhenon) [inline src/south/OPCUA/OPCUA_HA/OPCUA_HA.class.js:105]: logger.debug?

**Pull request merged**